src/util/permissions.tsx

```diff
@@ -0,0 +1,359 @@
+
+    if (processedResources.has(resourceLabel)) {
+      continue;
+    }
+
+    processedResources.add(resourceLabel);
+    resourceOptions.push({
+      value: permission.url,
+      label: resourceLabel,
+    });
+  }
+
+  resourceOptions.sort(sortOptions);
+
+  return resourceOptions;
+};
+
+export const generateResourceEntitlementOptions = (
+  resourceType: string,
+  permissions?: LxdPermission[],
+) => {
+  if (!permissions || !resourceType) {
+    return {};
+  }
+
+  const resourceEntitlements: Record<
+    string,
+    OptionHTMLAttributes<HTMLOptionElement>[]
+  > = {};
+
+  for (const permission of permissions) {
+    const resource = permission.url;
+    if (!resourceEntitlements[resource]) {
+      resourceEntitlements[resource] = [defaultOption];
+    }
+
+    resourceEntitlements[resource].push({
+      value: permission.entitlement,
+      label: permission.entitlement,
+    });
+  }
+
+  for (const resource in resourceEntitlements) {
+    // split entitlement options into two based on 'can_' prefix
+    const genericEntitlements: OptionHTMLAttributes<HTMLOptionElement>[] = [];
+    const granularEntitlements: OptionHTMLAttributes<HTMLOptionElement>[] = [];
+
+    for (const entitlementOption of resourceEntitlements[resource]) {
+      const entitlementValue = entitlementOption.value as string;
+      if (entitlementValue.includes("can_")) {
+        granularEntitlements.push(entitlementOption);
+        continue;
+      }
+
+      genericEntitlements.push(entitlementOption);
+    }
+
+    genericEntitlements.sort(sortOptions);
+    granularEntitlements.sort(sortOptions);
+
+    // add disabled option as a delimiter between the two sets of entitlements
+    if (genericEntitlements.length > 1 && granularEntitlements.length) {
+      genericEntitlements.splice(1, 0, {
+        disabled: true,
+        label: "Built-in entitlements",
+        value: "",
+      });
+
+      granularEntitlements.unshift({
+        disabled: true,
+        label: "Granular entitlements",
+        value: "",
+      });
+    }
+    resourceEntitlements[resource] =
+      genericEntitlements.concat(granularEntitlements);
+  }
+
+  return resourceEntitlements;
+};
+
+// refer to api spec to see how the names can be extracted from resource url
+// https://documentation.ubuntu.com/lxd/en/latest/api/
+export const extractResourceDetailsFromUrl = (args: {
+  resourceType: string;
+  path: string;
+  imageNamesLookup?: Record<string, string>;
+  identityNamesLookup?: Record<string, string>;
+}): ResourceDetail => {
+  const {
+    resourceType,
+    path,
+    imageNamesLookup = {},
+    identityNamesLookup = {},
+  } = args;
+  const url = new URL(`http://localhost/${path}`);
+  const project = url.searchParams.get("project");
+  const target = url.searchParams.get("target");
+  const urlSegments = path.split("/");
+  const name = decodeURIComponent(urlSegments[urlSegments.length - 1]).split(
+    "?",
+  )[0];
```

Comment:
the split seems not needed as the path won't contain a ?

Reply:
Aaah yes correct, I wasn't using `url.pathname` for setting up `urlSegments`, so had to do the split but that's not needed :+1: